components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/artifact/SAMLSSOArtifactResolutionService.java

```diff
@@ -116,29 +132,59 @@ public ArtifactResponse sendArtifactResolveRequest(ArtifactResolve artifactResol
-        if (matcher.find()) {
-            String artifactResponseReceived = matcher.group(0);
-            if (log.isDebugEnabled()) {
-                log.debug("Artifact Response: " + artifactResponseReceived);
-            }
-            try {
-                ArtifactResponse artifactResponse = (ArtifactResponse) SSOUtils.unmarshall(artifactResponseReceived);
-                if (isArtifactResponseValid(artifactResolve, artifactResponse)) {
-                    return artifactResponse;
+        if (artifactResponseString == null) {
+            throw new ArtifactResolutionException("SOAP response was null when resolving artifact: " +
+            artifactResolve.getArtifact().getArtifact() + " Issuer: " + artifactResolve.getIssuer().getValue());
+        }
+        ArtifactResponse artifactResponse = extractArtifactResponse(artifactResponseString);
+
+        validateArtifactResponse(artifactResolve, artifactResponse);
+        return artifactResponse;
+    }
+
+    /**
+     * Extract Artifact response object from soap message string. Return null if fail.
+     *
+     * @param artifactResponseString Response string from artifact resolver.
+     * @return Extracted artifact response object.
+     * @throws ArtifactResolutionException
+     */
+    public ArtifactResponse extractArtifactResponse(String artifactResponseString)
+            throws ArtifactResolutionException {
+
+        ArtifactResponse artifactResponse = null;
+        InputStream stream = new ByteArrayInputStream(artifactResponseString.getBytes(StandardCharsets.UTF_8));
+        try {
+            MessageFactory messageFactory = MessageFactory.newInstance();
+            SOAPMessage soapMessage = messageFactory.createMessage(new MimeHeaders(), stream);
+            SOAPBody soapBody = soapMessage.getSOAPBody();
+            Iterator iterator = soapBody.getChildElements();
+
+            while (iterator.hasNext()) {
+                SOAPBodyElement artifactResponseElement = (SOAPBodyElement) iterator.next();
+
+                if (StringUtils.equals(SAMLConstants.SAML20P_NS, artifactResponseElement.getNamespaceURI()) &&
+                        StringUtils.equals(ArtifactResponse.DEFAULT_ELEMENT_LOCAL_NAME,
+                                artifactResponseElement.getLocalName())) {
+
+                    DOMSource source = new DOMSource(artifactResponseElement);
+                    StringWriter stringResult = new StringWriter();
+                    TransformerFactory.newInstance().newTransformer().transform(
+                            source, new StreamResult(stringResult));
+                    artifactResponse = (ArtifactResponse) SSOUtils.unmarshall(stringResult.toString());
                 } else {
-                    throw new ArtifactResolutionException("Artifact Response is not valid.");
+                    throw new ArtifactResolutionException("Didn't receive valid artifact response.");
                 }
```

Comment:
So based on this if else block if the very first element is not an artifact response an error will be thown.
Is this the expectation. Can artifact response be nested ?

Reply:
According to the spec, the SOAP message should wrap the Artifact Response object. Following is an example given in the spec.

```
<SOAP-ENV:Envelope
    xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/">
    <SOAP-ENV:Body>
        <samlp:ArtifactResponse
            xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol"
            xmlns="urn:oasis:names:tc:SAML:2.0:assertion"
            ID="_FQvGknDfws2Z" Version="2.0"
            InResponseTo="_6c3a4f8b9c2d"
            IssueInstant="2004-01-21T19:00:49Z">
            <Issuer>https://wso2is.com</Issuer>
            <samlp:Status>
                <samlp:StatusCode Value="urn:oasis:names:tc:SAML:2.0:status:Success"/>
            </samlp:Status>
            <samlp:Response ID="d2b7c388cec36fa7c39c28fd298644a8"
                IssueInstant="2004-01-21T19:00:49Z"
                Version="2.0">
                ...
            </samlp:Response>
        </samlp:ArtifactResponse>
    </SOAP-ENV:Body>
</SOAP-ENV:Envelope>
```